### PR TITLE
acceptance: add testconfig and get it working with localcluster

### PR DESCRIPTION
**Pull request by BramGruneir**
This is the first step in getting more robust testing scenarios into 

I'm not wedded to the TestConfig name.  If you can think of something better I'd love to change it.

Follow ups to this will be:
1 - Getting the test configs working on remote acceptance tests
2 - Adding in a size flag so when setting up a store you can specify the size (and include this in the tests)
3 - Adding in all the chaos monkey agents.
4 - A larger collection of default tests.
5 - Consider having more of these tests run in circle and nightly jobs

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4446)

### CODE REVIEW

**Comment by vivekmenezes** [comment]
Review status: 0 of 11 files reviewed at latest revision, 11 unresolved discussions.

---

*[acceptance/cluster/localcluster.go, line 316 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkzCYVcdbqoGAp8WL4:-KAkzCYVcdbqoGAp8WL5:-1300603057) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/cluster/localcluster.go#L316)):*
// Expand the configuration into nodes and stores per node.

---

*[acceptance/cluster/localcluster.go, line 318 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkytTF1rYMHOdh2no0:-KAkytTF1rYMHOdh2no1:2127820678) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/cluster/localcluster.go#L318)):*
move storeCount further down

---

*[acceptance/cluster/localcluster.go, line 324 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkzUmx1b0w165bHm1I:-KAkzUmx1b0w165bHm1J:-211270194) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/cluster/localcluster.go#L324)):*
Is nodeCount == 0 valid? If you increment here you start off at 1.

---

*[acceptance/cluster/localcluster.go, line 435 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkzxmsZe59FNaOgU7c:-KAkzxmsZe59FNaOgU7d:1329174468) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/cluster/localcluster.go#L435)):*
No comma?

---

*[acceptance/main_test.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAl-oi86d9Eg5C5QK7V:-KAl-oi86d9Eg5C5QK7W:40335519) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/main_test.go#L57)):*
Are the two flags still needed? Or should we use the flags to change the defaults used to generate the config? Perhaps we should structure this in many layers. the flags are used to configure the configuration, which also can be modified by adding more shapes to the default configuration.

---

*[acceptance/single_key_test.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAl0wpM5r__H-MVI50u:-KAl0wpM5r__H-MVI50v:91012301) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/single_key_test.go#L38)):*
perhaps we add a new interface func(t *testing.T, cfg *ClusterConfig) that is run against all the configs so we don't have to repeat the loop around the cluster configs. The sql guys did this for their benchmarking code. 

---

*[acceptance/testconfig/testconfig.go, line 17 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkwJBKsq59kQ95GuVr:-KAkwJBKsq59kQ95GuVs:1355150456) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/testconfig/testconfig.go#L17)):*
I think this code and the proto can be moved to acceptance/cluster

---

*[acceptance/testconfig/testconfig.go, line 47 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkwzBjZBEXQyvtl2aR:-KAkwzBjZBEXQyvtl2aS:1195001367) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/testconfig/testconfig.go#L47)):*
I'd also add these in using code. You just need to take a default ClusterConfig and keep changing its name and Count.

---

*[acceptance/testconfig/testconfig.go, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkxguZB8rzodYve4J8:-KAkxguZB8rzodYve4J9:-994112594) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/testconfig/testconfig.go#L74)):*
I actually like the compressed config representation provided by the protobuf. Wanna hold off using this expanded form?

---

*[acceptance/testconfig/testconfig.proto, line 35 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkvvZwA7EZXOlXYov6:-KAkvvZwA7EZXOlXYov7:-222932408) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/testconfig/testconfig.proto#L35)):*
ClusterConfig?

---

*[acceptance/testconfig/testconfig.proto, line 39 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkuiBS5QXdlS3QrTme:-KAkuiBS5QXdlS3QrTmf:-1776554680) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/testconfig/testconfig.proto#L39)):*
Can you reorder these. I like name to be at the top.

Also can you add a comment above duration and stall?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446)*

**Comment by bdarnell** [comment]
Review status: 0 of 11 files reviewed at latest revision, 12 unresolved discussions.

---

*[acceptance/gossip_peerings_test.go, line 165 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAlXZQBCztGXKZHAyUR:-KAlXZQBCztGXKZHAyUS:1287802124) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/gossip_peerings_test.go#L165)):*
Why does this ignore the passed-in `clusterConfig`?  Why even run this test once per config?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446)*

**Comment by BramGruneir** [comment]
Review status: 0 of 11 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[acceptance/cluster/localcluster.go, line 316 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkzCYVcdbqoGAp8WL4:-KApBj1e7fzNqZ03dAJw:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/cluster/localcluster.go#L316)):*
Done.

---

*[acceptance/cluster/localcluster.go, line 318 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkytTF1rYMHOdh2no0:-KApBt3qg7FDXQdtr91O:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/cluster/localcluster.go#L318)):*
Done.

---

*[acceptance/cluster/localcluster.go, line 324 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkzUmx1b0w165bHm1I:-KApBvgBpN6Ec1WdCC34:-1872469694) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/cluster/localcluster.go#L324)):*
0 shouldn't be valid.  But I don't see why that is a problem here.  You won't get into the loop at all if that's the case.

---

*[acceptance/cluster/localcluster.go, line 435 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkzxmsZe59FNaOgU7c:-KApCRzGCSceXC-sMGiC:1155850567) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/cluster/localcluster.go#L435)):*
Ah, good call.  Fixed.

---

*[acceptance/gossip_peerings_test.go, line 165 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAlXZQBCztGXKZHAyUR:-KApURTvtc82lZ-eaOZl:-1615311215) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/gossip_peerings_test.go#L165)):*
Ah, it should use the configuration.  As to why run this test more than once, it is worthwhile to run it for different number of nodes.  But you're right, the other configurations don't really add anything.  I'll put a todo in to only pass in a subset of node sizes.

---

*[acceptance/main_test.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAl-oi86d9Eg5C5QK7V:-KApVMaQp5t-7iIHTstg:1564404462) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/main_test.go#L57)):*
I was planning on removing the flags once I add the ability to pass in a config file or a single line config.

---

*[acceptance/single_key_test.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAl0wpM5r__H-MVI50u:-KAq0HoUydo7ORfOsACB:-1203367924) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/single_key_test.go#L38)):*
Can you point me to an example?  How would this work?

---

*[acceptance/testconfig/testconfig.go, line 17 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkwJBKsq59kQ95GuVr:-KAlQjSUwaV9S_6NAlgn:2129167674) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/testconfig/testconfig.go#L17)):*
Good call.  Done.

---

*[acceptance/testconfig/testconfig.go, line 47 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkwzBjZBEXQyvtl2aR:-KAq0SZ1kUB3Yf06V7IS:841410415) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/testconfig/testconfig.go#L47)):*
I'm going to eventually move these to a file.  So automation seemed premature.

---

*[acceptance/testconfig/testconfig.go, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkxguZB8rzodYve4J8:-KAlR8nUxjEjvTlnIS-7:-1970545917) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/testconfig/testconfig.go#L74)):*
Really?  I find the JSON style hard to parse quickly.  But I'm cool with dropping this code.

---

*[acceptance/testconfig/testconfig.proto, line 35 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkvvZwA7EZXOlXYov6:-KAlRN2-odQoYNI1VEwf:296778702) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/testconfig/testconfig.proto#L35)):*
If it was only the cluster, than sure.  But it's got a bunch of test params in there too.   Now that it's in the cluster package, I think testconfig works.  Since it is now Cluster.TestConfig

---

*[acceptance/testconfig/testconfig.proto, line 39 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkuiBS5QXdlS3QrTme:-KAq2zC7pgp0rd8pXQ9q:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/testconfig/testconfig.proto#L39)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446)*

**Comment by vivekmenezes** [comment]
LGTM

---

Review status: 0 of 11 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[acceptance/cluster/localcluster.go, line 324 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkzUmx1b0w165bHm1I:-KArNTsBCCwn0iKSRd7d:1669848617) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/cluster/localcluster.go#L324)):*
I noticed that you were incrementing storeCount at the end of stores loop, so I was puzzled why you were pre-incrementing the node count.

---

*[acceptance/cluster/localcluster.go, line 109 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KArMfYSlMP15-kQB4I7:-KArMfYSlMP15-kQB4I8:43322377) ([raw file](https://github.com/cockroachdb/cockroach/blob/395f9e553163335a6d612db7cee4551b872bb86f/acceptance/cluster/localcluster.go#L109)):*
Do you really need this index?

---

*[acceptance/cluster/localcluster.go, line 110 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KArMaEyfPHqWhGsBxU3:-KArMaEyfPHqWhGsBxU4:-1356023849) ([raw file](https://github.com/cockroachdb/cockroach/blob/395f9e553163335a6d612db7cee4551b872bb86f/acceptance/cluster/localcluster.go#L110)):*
is this the name? I also like the name to be the first field. same in testNode.

---

*[acceptance/main_test.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAl-oi86d9Eg5C5QK7V:-KArPMTHTDK62ceMcb66:-1410477852) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/main_test.go#L57)):*
Can you create an issue for that.

---

*[acceptance/main_test.go, line 64 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KArPSNDiFX2krILHShL:-KArPSNDiFX2krILHShM:-837385057) ([raw file](https://github.com/cockroachdb/cockroach/blob/395f9e553163335a6d612db7cee4551b872bb86f/acceptance/main_test.go#L64)):*
remove when you add that issue.

---

*[acceptance/single_key_test.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAl0wpM5r__H-MVI50u:-KArQLdGJTAbbDz6y_Zz:186935143) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/single_key_test.go#L38)):*
func TestSingleKey(t * testing.T) {
    cluster.RunOnAllTestClusters(t, testSingleKeyInner)
}

func RunOnAllConfigs(t *testing.T, testFunc func(t *testing.T, cfg cluster.TestConfig) {
       clusterConfigs := getConfigs()
	for _, clusterConfig := range clusterConfigs {
		testFunc(t, clusterConfig)
        }
}

---

*[acceptance/status_server_test.go, line 100 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KArRdm5K0nIlsMCK54h:-KArRdm5K0nIlsMCK54i:1676911776) ([raw file](https://github.com/cockroachdb/cockroach/blob/395f9e553163335a6d612db7cee4551b872bb86f/acceptance/status_server_test.go#L100)):*
use RunOnAllTestClusters

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-:-KArRjdm-ykIcEuEYahV:676834233)*

**Comment by BramGruneir** [comment]
Review status: 0 of 11 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[acceptance/cluster/localcluster.go, line 324 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAkzUmx1b0w165bHm1I:-KAsAqhDc0sxhgyns4zz:-1705609021) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/cluster/localcluster.go#L324)):*
Both are incremented right after they're used to create a new node or store.  But I can move it below the node append instead.  Just seemed more natural to me.

---

*[acceptance/cluster/localcluster.go, line 109 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KArMfYSlMP15-kQB4I7:-KAsBlyNcqFgj2Xcj7ie:380002936) ([raw file](https://github.com/cockroachdb/cockroach/blob/395f9e553163335a6d612db7cee4551b872bb86f/acceptance/cluster/localcluster.go#L109)):*
It's used to create the dataStr and I thought it might be useful when we start killing single stores. 

---

*[acceptance/cluster/localcluster.go, line 110 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KArMaEyfPHqWhGsBxU3:-KAsAZfaguU5dQbJxMvn:-110319134) ([raw file](https://github.com/cockroachdb/cockroach/blob/395f9e553163335a6d612db7cee4551b872bb86f/acceptance/cluster/localcluster.go#L110)):*
No, not the name.  This is the string used when creating the teststore.  dataStr(nodeCount, storeCount)  This is the string used when starting the cockroach cluster/container.  I've renamed them to dataStr and nodeStr which match the commands that create them.

---

*[acceptance/main_test.go, line 57 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAl-oi86d9Eg5C5QK7V:-KAs1goF1ySuWOs2fG28:-1002324739) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/main_test.go#L57)):*
Sure, done #4509

---

*[acceptance/main_test.go, line 64 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KArPSNDiFX2krILHShL:-KAs2oOlt5HzZKOKxPFh:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/395f9e553163335a6d612db7cee4551b872bb86f/acceptance/main_test.go#L64)):*
Done.

---

*[acceptance/single_key_test.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KAl0wpM5r__H-MVI50u:-KAs2vRiRW3jgCCygOyK:86007975) ([raw file](https://github.com/cockroachdb/cockroach/blob/307845a378d45cc27c85e3e3415f56a19c9d79c5/acceptance/single_key_test.go#L38)):*
Ah, I see what you mean.  Sure, I'm always up for some DRYer code.  I even moved the cluster creation  and assertandstop into it.

---

*[acceptance/status_server_test.go, line 100 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-KArRdm5K0nIlsMCK54h:-KAs6_-NjGY6wi4lG0ge:-1707490625) ([raw file](https://github.com/cockroachdb/cockroach/blob/395f9e553163335a6d612db7cee4551b872bb86f/acceptance/status_server_test.go#L100)):*
Done.  And all the other tests as well.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446)*

**Comment by BramGruneir** [comment]
Update, rebased, fixedup and pushed.  Please give it one final look before I merge.

---

Review status: 0 of 11 files reviewed at latest revision, 16 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4446#-:-KAsdg1JpQpfzmFfTZXx:969211396)*
